src/tools/dynamic/sf-list-tools.ts

```diff
@@ -30,14 +40,46 @@ ONLY use this tool if:
 2. You genuinely don't know what tools are available for a specific task
 3. You need to discover new tools for an unfamiliar use case
 
+When calling this tool, provide a description of what you're trying to accomplish in the query parameter.
+Examples: "deploy a file to Salesforce org", "create a scratch org", "retrieve metadata from org"
+
 If you find one or more tools you want to enable, call sf-enable-tools with all the tool names.
 Once you have enabled a tool, you MUST invoke the tool to accomplish the user's original request - DO NOT USE A DIFFERENT TOOL OR THE COMMAND LINE.
 Once a tool has been enabled, you do not need to call sf-list-tools again - instead, invoke the desired tool directly.`,
+    listToolsParamsSchema.shape,
     {
       title: 'List all individual tools',
       readOnlyHint: true,
       openWorldHint: false,
     },
-    async () => textResponse(JSON.stringify(await listAllTools(), null, 2))
+    async ({ query }) => {
+      const assets = await getToolSearchAssets();
+
+      // Embed the user query
+      const queryEmbedding = await assets.embedder(query, {
+        pooling: 'mean',
+        normalize: true,
+      });
+
+      // Perform Semantic Search (FAISS)
+      const searchResults = assets.faissIndex.search(
+        // Convert the embedding tensor data to a flat array of numbers
+        Array.from(queryEmbedding.data as Float32Array),
+        5
+      );
+
+      const allTools = await listAllTools();
+      const enabledToolNames = new Set(allTools.filter((tool) => tool.enabled).map((tool) => tool.name));
+
+      // Filter search results to exclude enabled tools and get top 5
+      const filteredResults = searchResults.labels
+        .map((id) => assets.tools.find((c) => c.id === id)!)
+        .filter((tool) => !enabledToolNames.has(tool.name))
```

Comment:
This `!enabledToolName` might cause some trouble... Check out this video, it all looked like it was going well until the model chose to run `sf-list-all-tools` a second time. Since the result filters out enabled tools, the `sf-query-org` tool was not returned the second time and it chose to build an sf command instead. 

In the second call the query was: 
```
{
  "query": "query SOQL records from Salesforce org"
}
```

Maybe this can be fixed by tweaking the description? Something like: "Don't run this tool twice for the same basic query." or "Always check enabled tools before running `sf-list-all-tools`?

https://github.com/user-attachments/assets/851037c6-5d5f-4e8a-a811-1e3c67219bff